scripts/referrer-revenue-summary.ts

```diff
@@ -0,0 +1,161 @@
+      type: 'number',
+      demandOption: false,
+      default: 0,
+    })
+    .option('end-timestamp', {
+      alias: 'e',
+      description:
+        'timestamp at which to stop checking for revenue calculation (since epoch)',
+      type: 'number',
+      demandOption: false,
+      default: Date.now(),
+    })
+    .option('referrer-ids', {
+      alias: 'r',
+      description: 'a comma separated list of referrers IDs',
+      type: 'string',
+      demandOption: false,
+      coerce: (arg) => new Set(arg.split(',')),
+    })
+    .option('network-ids', {
+      alias: 'n',
+      description: 'comma-separated list of network IDs',
+      type: 'string',
+      demandOption: false,
+      coerce: (arg: string) => new Set(arg.split(',')),
+    }).argv
+
+  return {
+    protocolId: argv['protocol'] as Protocol,
+    protocolFilter: protocolFilters[argv['protocol'] as Protocol],
+    startTimestamp: argv['start-timestamp'],
+    endTimestamp: argv['end-timestamp'],
+    networkIds: (argv['network-ids'] as NetworkId) ?? supportedNetworkIds,
+    referrers: argv['referrer-ids'] as string,
+    outputFile: argv['output-file'],
+  }
+}
+
+async function main() {
+  const args = await getArgs()
+  const startTimestamp = new Date(args.startTimestamp)
+  const endTimestamp = new Date(args.endTimestamp)
+  const networkIdsArray = [
+    ...(args.networkIds ?? supportedNetworkIds),
+  ] as NetworkId[]
+  const referrerArray = args.referrers ? [...args.referrers] : undefined
+
+  const referralEvents = await fetchReferralEvents(
+    networkIdsArray,
+    args.protocolId,
+    referrerArray,
+  )
+
+  const uniqueEvents = removeDuplicates(referralEvents)
+  const protocolFilteredEvents: ReferralEvent[] =
+    await args.protocolFilter(uniqueEvents)
+  const revenueCalcHandler = calculateRevenueHandlers[args.protocolId]
+
+  interface ReferralInfo {
+    totalRevenue: number
+    referralCount: number
+    averageRevenue?: number
+  }
+
+  const referrerMap = new Map<string, ReferralInfo>()
+
+  for (const event of protocolFilteredEvents) {
+    const revenue = await revenueCalcHandler({
+      address: event.userAddress,
+      startTimestamp,
+      endTimestamp,
+    })
+
+    const referrerInfo = referrerMap.get(event.referrerId)
+    if (referrerInfo) {
+      referrerInfo.totalRevenue += revenue
+      referrerInfo.referralCount++
+    } else {
+      referrerMap.set(event.referrerId, {
+        totalRevenue: revenue,
+        referralCount: 1,
+      })
+    }
+  }
+
+  // Calculate average revenue done here to avoid calculating at each iteration
+  for (const [_, referrerInfo] of referrerMap) {
+    referrerInfo.averageRevenue =
+      referrerInfo.totalRevenue / referrerInfo.referralCount || 0
+  }
+
+  const allResultsArray: [string, number, number, number | undefined][] = []
+
+  // Handle cases were a cli passed referrer ID param that has no results
+  if (referrerArray) {
```

Comment:
why don't you do this first and unconditionally (if there's not a list of referrers passed in, get the list).  It will clean up the condition about around "if the thing doesn't exist yet, set it to 1"

Reply:
Changed in 08e3adb6bd5e5bc65908f80b9fdce134bdd4979c!